python/ray/serve/backend_worker.py

```diff
@@ -1,21 +1,50 @@
+import asyncio
 import traceback
 import inspect
 from collections.abc import Iterable
+from collections import defaultdict
+from itertools import groupby
+from operator import attrgetter
 
 import ray
+from ray.async_compat import sync_to_async
+
 from ray import serve
 from ray.serve import context as serve_context
 from ray.serve.context import FakeFlaskRequest
-from collections import defaultdict
 from ray.serve.utils import (parse_request_item, _get_logger,
-                             retry_actor_failures)
+                             retry_actor_failures, chain_future, unpack_future)
 from ray.serve.exceptions import RayServeException
 from ray.serve.metric import MetricClient
-from ray.async_compat import sync_to_async
+from ray.serve.config import BackendConfig
+from ray.serve.router import Query
 
 logger = _get_logger()
 
 
+class WaitableQueue(asyncio.Queue):
+    async def wait_for_batch(self, num_items: int, timeout_s: float):
+        """Wait up to num_items in the queue given timeout_s."""
+
+        assert num_items >= 1
+        # Wait for the first value without timeout. We will return at least
+        # one item. Additionally this help the caller context switch on empty
+        # queue.
+        batch = [
+            await self.get(),
+        ]
+
+        # Wait for the remaining batch with the timeout
+        if num_items > 1:
+            done_set, not_done_set = await asyncio.wait(
+                [self.get() for _ in range(num_items - 1)], timeout=timeout_s)
```

Comment:
Shouldn't the timeout clock start at the beginning of the function call? For example if waiting for the first call takes longer than the timeout, we should return immediately after that.

Reply:
I considered that. The issue here is that the `main_loop` quickly becomes a busy spinning loop when timeout=0.

Reply:
Can't you just call `asyncio.wait` with timeout 0 once the timeout is up?

Reply:
asycio.wait with timeout 0 will returns almost immediately.

I was talking about the `main_loop` implementation:
```
def main_loop:
    while True:
        pop from queue
        run batch
```
if the pop from queue keep returning zero items, this loop will be a busy loop.

Reply:
Right, but we're blocking for the first item anyways so there shouldn't be a busy spin, right? The current behavior doesn't make any sense to me because we block indefinitely for the first item and then block for the timeout waiting for further items if I'm understanding this correctly.